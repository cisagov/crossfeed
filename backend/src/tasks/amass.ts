import { Domain } from '../models';
import { spawnSync } from 'child_process';
import { readFileSync } from 'fs';
import { plainToClass } from 'class-transformer';
import { CommandOptions } from './ecs-client';
import getRootDomains from './helpers/getRootDomains';
import saveDomainsToDb from './helpers/saveDomainsToDb';
import * as path from 'path';

const OUT_PATH = path.join(__dirname, 'out-' + Math.random() + '.txt');

export const handler = async (commandOptions: CommandOptions) => {
  const { organizationId, organizationName, scanId } = commandOptions;

  console.log('Running amass on organization', organizationName);

  const rootDomains = await getRootDomains(organizationId!);

  for (const rootDomain of rootDomains) {
    try {
      const args = [
        'enum',
        '-ip',
        '-active',
        '-d',
        rootDomain,
        '-json',
        OUT_PATH
      ];
      console.log('Running amass with args', args);
      spawnSync('amass', args, { stdio: 'pipe' });
      const output = String(readFileSync(OUT_PATH));
      const lines = output.split('\n');
      const domains: Domain[] = [];
      for (const line of lines) {
        if (line == '') continue;
        const parsed = JSON.parse(line);
        domains.push(
          plainToClass(Domain, {
            ip: parsed.addresses[0].ip,
            name: parsed.name,
            asn: parsed.addresses[0].asn,
            organization: { id: organizationId },
            fromRootDomain: rootDomain,
            discoveredBy: { id: scanId }
          })
        );
      }
      await saveDomainsToDb(domains);
      console.log(`amass created/updated ${domains.length} new domains`);
    } catch (e) {
      console.error(e);
      continue;
    }
  }
};
