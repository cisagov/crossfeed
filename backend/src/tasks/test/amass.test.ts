import { handler as amass } from '../amass';
jest.mock('../helpers/getRootDomains');
jest.mock('../helpers/saveDomainsToDb');

jest.mock('fs', () => ({
  readFileSync: () => `{"name":"filedrop.cisa.gov","domain":"cisa.gov","addresses":[{"ip":"2a02:26f0:7a00:195::447a","cidr":"2a02:26f0:7a00::/48","asn":6762,"desc":"SEABONE-NET TELECOM ITALIA SPARKLE S.p.A."},{"ip":"2a02:26f0:7a00:188::447a","cidr":"2a02:26f0:7a00::/48","asn":6762,"desc":"SEABONE-NET TELECOM ITALIA SPARKLE S.p.A."}],"tag":"dns","source":"DNS"}
{"name":"www.filedrop.cisa.gov","domain":"cisa.gov","addresses":[{"ip":"2a02:26f0:7a00:188::447a","cidr":"2a02:26f0:7a00::/48","asn":6762,"desc":"SEABONE-NET TELECOM ITALIA SPARKLE S.p.A."},{"ip":"2a02:26f0:7a00:195::447a","cidr":"2a02:26f0:7a00::/48","asn":6762,"desc":"SEABONE-NET TELECOM ITALIA SPARKLE S.p.A."}],"tag":"cert","source":"Crtsh"}`,
  unlinkSync: () => null
}));
jest.mock('child_process', () => ({
  spawnSync: () => null
}));

describe('amass', () => {
  test('basic test', async () => {
    await amass({
      organizationId: 'organizationId',
      organizationName: 'organizationName',
      scanId: 'ee7640ae-62d9-4e01-8a32-0e42dc8365dd',
      scanName: 'scanName',
      scanTaskId: 'scanTaskId'
    });
  });
});
