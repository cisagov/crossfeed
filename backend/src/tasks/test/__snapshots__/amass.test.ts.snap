// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`amass basic test: helpers.saveDomainsToDb 1`] = `
Array [
  Domain {
    "asn": 6762,
    "discoveredBy": Object {
      "id": "ee7640ae-62d9-4e01-8a32-0e42dc8365dd",
    },
    "fromRootDomain": "filedrop.cisa.gov",
    "ip": "2a02:26f0:7a00:195::447a",
    "name": "filedrop.cisa.gov",
    "organization": Object {
      "id": "organizationId",
    },
  },
  Domain {
    "asn": 6762,
    "discoveredBy": Object {
      "id": "ee7640ae-62d9-4e01-8a32-0e42dc8365dd",
    },
    "fromRootDomain": "filedrop.cisa.gov",
    "ip": "2a02:26f0:7a00:188::447a",
    "name": "www.filedrop.cisa.gov",
    "organization": Object {
      "id": "organizationId",
    },
  },
]
`;
